### ISSUE TITLE
Can diversion sessions have system wide side effects?

### ISSUE THREAD

**Issue by sidkshatriya** [mozilla/rr#1772]
Lets say I do the following during replay:

```
gdb > call someFunction()
```

This would execute `someFunction()` in a fork as I understand it.

Now typically `someFunction()` will output something to stdout or can have some return value I'm interested in (in that case I would `p someFunction()` instead).

What if `someFunction()` does something else -- like write to a database? I'm guessing since the database is probably an emulated socket/device during replay it has no real effect? But then how do you allow writes to stdout during diversion calls (is that an exception) ?

Basically do my diversion sessions have the ability to affect the computer state (databases, disks, network) (which a replay session does not)? Related to this... how exactly does a diversion session run in fork? Are all syscalls "real" syscalls now?

**Comment by Keno** [comment]
Yes, all syscalls are real, with the exception of those listed here: https://github.com/mozilla/rr/blob/master/src/DiversionSession.cc#L62, though of course as you noted some file descriptors are emulated those syscalls will fail.

**Comment by sidkshatriya** [comment]
Thanks for your reply. So just as some file descriptors are emulated (during diversion), the corollary to that is some are "real" e.g. stdin, stdout, stderr -- correct?

**Comment by rocallahan** [comment]
Correct.
